SimTracker/SiPixelDigitizer/plugins/SiPixelChargeReweightingAlgorithm.cc

```diff
@@ -230,8 +230,11 @@ bool SiPixelChargeReweightingAlgorithm::hitSignalReweight(const PSimHit& hit,
   for (int row = 0; row < TXSIZE; ++row) {
     for (int col = 0; col < TYSIZE; ++col) {
       //Fill charges into 21x13 Pixel Array with hitPixel in centre
-      pixrewgt[row][col] =
-          hitSignal[PixelDigi::pixelToChannel(hitPixel.first + row - THX, hitPixel.second + col - THY)];
+      if ((hitPixel.first + row - THX) >= 0 && (hitPixel.first + row - THX) < topol->nrows() &&
```

Comment:
instead of four comparisons for each iteration would not be more efficient to change the loop boundaries?

Reply:
@VinInn, thanks for the comment. What is done here is essentially a copy of what is done in the same code a few dozen lines below, namely https://github.com/cms-sw/cmssw/blob/CMSSW_12_1_0_pre3/SimTracker/SiPixelDigitizer/plugins/SiPixelChargeReweightingAlgorithm.cc#L273-L274, which, I agree now that you point it out, is not the most efficient approach from the performance point of view.

It seems that this 21x13 template array can occasionally go across the physical boundaries of the pixel module and while a check of this is done after the charge reweighting, the same check has not been done before the reweighting when charges are being filled into the template array.

It would therefore be best to update the loop boundaries in both places. I will look into this.